### DISCUSSION TITLE
Directory Structure for App with Protected Content

### DISCUSSION THREAD

**Discussion by nvijayap** [nuxt/nuxt#24121] [Questions] [answered]
Hi,

I am embarking on creating a Nuxt 3 app. I would like to know how to layout the directory structure. The app has a landing page (unprotected) and a bunch of protected (secure) pages that are rendered/visible only after user authenticates to an external IAM/SSO server. How to go about? 

Thanks

**Comment by TheAlexLichter** [comment, marked as answer]
My suggestion: Move the content to protect behind an API and fetch it only if the user is authenticated.

The directory structure should follow the structure of your page (which you didn't explain here).

**Reply by nvijayap** [reply]
OK, Thank you!

**Reply by kevhof** [reply]
How do you go about it when the content to protect is fully styled, stateful and dependant on user interaction? Like when it's a full component basically.

**Comment by AloisSeckar** [comment]
Another option is to use [middleware](https://nuxt.com/docs/guide/directory-structure/middleware) and redirect to login screen whenever the user is not authenticated. Either define it via `definePageMeta` on each page you want to be auth protected or create `.global` middleware that runs for every page (and inside the handler you skip the few pages that should be public).

**Reply by TheAlexLichter** [reply]
Be aware that the content of your site is still "readable" if it doesn't come from an API because the JS can be fetched even though there might be a middleware in place.

**Reply by nvijayap** [reply]
OK, is there a readily usable (lib, plugin, middleware, component) that can handle this Tuple5 (as needed for the IAM/SSO) server/system ? ...

1. client_id
2. client_secret
3. username
4. password
5. grant_type

I am looking in the context of Nuxt3

**Reply by TheAlexLichter** [reply]
Never heard of Tuple5 unfortunately.

**Reply by nvijayap** [reply]
Sorry for using a terminology from Scala ... it is just indicative of any 5 items ...
```
scala> (1,2,3,4,5).getClass
val res3: Class[? <: (Int, Int, Int, Int, Int)] = class scala.Tuple5
```
So, all I am looking for is a library/component that can orchestrate with an IAM/SSO server that needs these 5 granular pieces.

**Reply by TheAlexLichter** [reply]
You can try sth. like [nuxt-auth from sidebase](https://github.com/sidebase/nuxt-auth) or [authjs-nuxt](https://github.com/Hebilicious/authjs-nuxt) - both supporting OAuth2 flows of various kinds

**Reply by nvijayap** [reply]
Excellent, Thank you!  I appreciate, that serves my purpose.

**Reply by kevhof** [reply]
Is this method really secure? The route protection happens client-side, so I guess a developer could easily overpass the protection right?

**Reply by TheAlexLichter** [reply]
> How do you go about it when the content to protect is fully styled, stateful and dependant on user interaction? Like when it's a full component basically.

Then you have no chance of "fully hiding it" because it will be part of the bundle.

**Reply by kevhof** [reply]
@manniL  That's what I thought... It's really a missing feature of Nuxt IMO. Being able to protect page routes **server-side**.

**Reply by TheAlexLichter** [reply]
@kevhof I don't think any framework that servers JS for the frontend can accomplish this. I've touched on this [on my recent video](https://www.youtube.com/watch?v=gMoipv_NkWE) ☺️

The best chance might be server components with slots, where the server component stores the "data" and re-uses existing "non-sensitive" components (e.g. buttons/links). That could work 🤔
